src/core/net/dhcp6_server.cpp

```diff
@@ -0,0 +1,544 @@
+                  //VerifyOrExit(((option.GetLength() == (sizeof(option) - sizeof(Dhcp6Option))) &&
+                  (option.GetDuidType() == kDuidLL) &&
+                  (HostSwap16(option.GetDuidLL().mDuidHardwareType) == HARDWARE_TYPE_EUI64) &&
+                  (!memcmp(option.GetDuidLL().mDuidAddrLL, mMac.GetExtAddress(), ADDRESS_LL_LEN))),
+                 error = kThreadError_Parse);
+exit:
+    return error;
+}
+
+ThreadError Dhcp6Server::ProcessIaNa(Message &aMessage, uint16_t aOffset,
+                                     Type aType, ClientIdentifier &aClient, IaNa &aIaNa, IaAddress &aIaAddress)
+{
+    ThreadError error = kThreadError_None;
+    uint16_t optionOffset;
+    uint16_t length;
+
+    VerifyOrExit((aMessage.Read(aOffset, sizeof(aIaNa), &aIaNa) == sizeof(aIaNa)), error = kThreadError_Parse);
+    aOffset += sizeof(aIaNa);
+    length = aIaNa.GetLength();
+    VerifyOrExit((optionOffset = FindOption(aMessage, aOffset, length, Dhcp6::kOptionIaAddress)) > 0, ;);
+    SuccessOrExit(error = ProcessIaAddress(aMessage, optionOffset, aType, aClient, aIaAddress));
+
+exit:
+    return error;
+}
+
+ThreadError Dhcp6Server::ProcessIaAddress(Message &aMessage, uint16_t aOffset, Type aType, ClientIdentifier &aClient,
+                                          IaAddress &aIaAddress)
+{
+    ThreadError error = kThreadError_None;
+    PrefixBuffer *prefix = mPrefixesHead;
+    AddressBuffer *address;
+
+    VerifyOrExit((aMessage.Read(aOffset, sizeof(aIaAddress), &aIaAddress) == sizeof(aIaAddress)),
+                 error = kThreadError_Parse);
+
+    switch (aType)
+    {
+    case kTypeSolicit:
+        while (prefix)
+        {
+            if (otIp6PrefixMatch(&aIaAddress.GetAddress(), &(prefix->GetPrefix().mPrefix)) >= prefix->GetPrefix().mLength)
+            {
+                address = prefix->GetAddressBuffer();
+
+                while (address)
+                {
+                    if (!memcmp(&(address->GetClient()), &aClient, sizeof(ClientIdentifier)))
+                    {
+                        memcpy(&aIaAddress, &(address->GetIaAddress()), sizeof(IaAddress));
+                        ExitNow();
+                    }
+
+                    address = address->GetNext();
+                }
+
+                address = mAddressesAvail;
+                mAddressesAvail = mAddressesAvail->GetNext();
+
+                memcpy(&(aIaAddress.GetAddress().mFields.m8[8]), aClient.GetDuidLL().mDuidAddrLL, 8);
+                address->SetClient(aClient);
+                address->SetIaAddress(aIaAddress);
+                address->SetNext(prefix->GetAddressBuffer());
+                prefix->SetAddressBuffer(address);
+                ExitNow();
+            }
+
+            prefix = prefix->GetNext();
+        }
+
+        break;
+
+    case kTypeRelease:
+        while (prefix)
+        {
+            if (otIp6PrefixMatch(&aIaAddress.GetAddress(), &(prefix->GetPrefix().mPrefix)) >= prefix->GetPrefix().mLength)
```

Comment:
I know you are trying to maintain a client entry within a prefix.  I think it would be simpler if you just maintained a single client table.  Each client entry maintains each address assigned to the client (one for each prefix).